File: backend/app/schemas/portal.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.photo import PhotoPose
from app.schemas.quote import QuoteOut


class PortalTokenOut(BaseModel):
    """What the trainer sees after issuing (or revoking) a portal link."""

    client_id: uuid.UUID
    portal_token: str | None
    portal_token_issued_at: datetime | None

    model_config = {"from_attributes": True}


class InviteTemplatesOut(BaseModel):
    """The trainer's wording, already resolved to the stock text where unset."""

    whatsapp_template: str
    email_subject: str
    email_template: str
    # Names the trainer may use between braces, so the UI can list them.
    placeholders: list[str]


class InviteTemplatesUpdate(BaseModel):
    """An empty field restores the stock text for that channel."""

    whatsapp_template: str | None = None
    email_subject: str | None = None
    email_template: str | None = None


class PortalInviteOut(BaseModel):
    """Ready-to-send wording for the portal link, in both channels."""

    url: str
    subject: str
    body: str
    whatsapp_text: str


class PortalWeighInCreate(BaseModel):
    """The only thing a client may write. The day is always today, never chosen."""

    weight_kg: float = Field(gt=0, le=500)


class PortalWeighInOut(BaseModel):
    """A weigh-in as the client sees it: no client id, and no trainer notes."""

    id: uuid.UUID
    measured_on: date
    weight_kg: float
    bmi: float | None


class PortalPhotoOut(BaseModel):
    """A progress photo as the client sees it: no client id, no path on disk.

    The id is only good for asking for the file through the same token, so it
    tells a leaked link nothing it could use anywhere else.
    """

    id: uuid.UUID
    taken_on: date
    pose: PhotoPose

    model_config = {"from_attributes": True}


class PortalClientOut(BaseModel):
    """What the client sees. Never carries the client id: the token is the key."""

    full_name: str
    goals: str | None
    latest_weight_kg: float | None
    latest_weighed_on: date | None
    # So the portal only offers what there is to open.
    has_training_plan: bool
    has_diet_plan: bool
    weigh_in_count: int
    # Null means consent was never given. The count is what makes "delete my
    # photos" honest: the client sees how many there are before deciding.
    photo_consent_at: datetime | None
    photo_count: int
    # How many sessions the client has logged, so the portal only offers the
    # progression screen once there is something to draw.
    workout_count: int
    # Today's message, pinned or drawn from the rotation. Null when the trainer
    # has not written any yet.
    quote: QuoteOut | None = None

File: frontend/src/types/portal.ts
import type { MacroTotals } from "./diet";
import type { PhotoPose } from "./photo";
import type { Quote } from "./quote";

export interface PortalToken {
  client_id: string;
  portal_token: string | null;
  portal_token_issued_at: string | null;
}

/** Ready-to-send wording for the link, composed by the backend. */
export interface PortalInvite {
  url: string;
  subject: string;
  body: string;
  whatsapp_text: string;
}

/** What the client sees through their own link. Carries no client id by design. */
export interface PortalHome {
  full_name: string;
  goals: string | null;
  latest_weight_kg: number | null;
  latest_weighed_on: string | null;
  has_training_plan: boolean;
  has_diet_plan: boolean;
  weigh_in_count: number;
  /** Null means the client never agreed to their progress photos being kept. */
  photo_consent_at: string | null;
  photo_count: number;
  /** Sessions logged: the portal only offers the charts once there are any. */
  workout_count: number;
  /** Today's message: the one pinned to this client, or one from the rotation. */
  quote: Quote | null;
}

/** A progress photo as the client gets it: no client id, no path on disk. */
export interface PortalPhoto {
  id: string;
  taken_on: string;
  pose: PhotoPose;
}

/**
 * The plans arrive already flattened by the backend — the very same document the
 * PDF is rendered from — so the portal shows exercise names and pictures with no
 * lookup of its own. The catalogue id rides along only so that tapping a row can
 * ask for its sheet, through the same token.
 */
export interface PortalExercise {
  /** The catalogue id, only good for opening the sheet through the same token. */
  exercise_id: string;
  name_es: string;
  sets: number;
  reps: string;
  rest_seconds: number | null;
  tempo: string | null;
  notes: string | null;
  image_path: string | null;
  /** "A1"/"A2" when it is chained to the next one, with no rest in between. */
  superset_label: string | null;
  /** The note about the block, carried by the exercise that opens it. */
  superset_note: string | null;
}

/** An exercise sheet as the client reads it: how it is done, and nothing else. */
export interface PortalExerciseDetail {
  id: string;
  name_es: string;
  category_es: string | null;
  level_es: string | null;
  force_es: string | null;
  mechanic_es: string | null;
  equipment_es: string | null;
  primary_muscles_es: string[];
  secondary_muscles_es: string[];
  instructions_es: string[];
  images: string[];
}

export interface PortalTrainingDay {
  day_of_week_es: string;
  exercises: PortalExercise[];
}

export interface PortalTrainingWeek {
  week_number: number;
  notes: string | null;
  days: PortalTrainingDay[];
}

export interface PortalTrainingPlan {
  client_name: string;
  plan_title: string;
  plan_notes: string | null;
  start_date: string | null;
  end_date: string | null;
  weeks: PortalTrainingWeek[];
}

export interface PortalMealItem {
  food_name: string;
  quantity_label: string | null;
  calories: number | null;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
}

export interface PortalMeal {
  name: string;
  time_of_day: string | null;
  items: PortalMealItem[];
  totals: MacroTotals;
}

export interface PortalDietDay {
  day_of_week_es: string;
  menu_name: string | null;
  meals: PortalMeal[];
  totals: MacroTotals | null;
}

export interface PortalDietWeek {
  week_number: number;
  notes: string | null;
  days: PortalDietDay[];
}

export interface PortalDietPlan {
  client_name: string;
  plan_title: string;
  plan_notes: string | null;
  start_date: string | null;
  end_date: string | null;
  daily_calories_target: number | null;
  daily_protein_g: number | null;
  daily_carbs_g: number | null;
  daily_fat_g: number | null;
  weeks: PortalDietWeek[];
}

export function portalPath(token: string): string {
  return `/p/${token}`;
}
